From 645a3b8abbb0c516fd1e2e63cd505e321bc14923 Mon Sep 17 00:00:00 2001
From: Olivier Brunel <jjk@jjacky.com>
Date: Fri, 17 Jan 2014 19:19:20 +0100
Subject: [PATCH 2/2] treeview: Fix visual glitch on row-inserted

With fixed row height, when rows are inserted but not visible there's
a call to queue_resize_no_redraw() to avoid useless redraw operation.

However, when calling gtk_tree_view_top_row_to_dy() to adjust the
adjustment/scroll position, a redraw would be performed leading to some
invalid visual "glitches" (e.g. rows moved to wrong place).

Disabling this when called from gtk_tree_view_top_row_to_dy() avoids such
issues.

https://bugzilla.gnome.org/show_bug.cgi?id=721597
Signed-off-by: Olivier Brunel <jjk@jjacky.com>
---
 gtk/gtktreeview.c | 26 ++++++++++++++++----------
 1 file changed, 16 insertions(+), 10 deletions(-)

diff --git a/gtk/gtktreeview.c b/gtk/gtktreeview.c
index fd65baf..c171bfe 100644
--- a/gtk/gtktreeview.c
+++ b/gtk/gtktreeview.c
@@ -11281,13 +11281,16 @@ gtk_tree_view_adjustment_changed (GtkAdjustment *adjustment,
   if (gtk_widget_get_realized (GTK_WIDGET (tree_view)))
     {
       gint dy;
-	
-      gdk_window_move (tree_view->priv->bin_window,
-		       - gtk_adjustment_get_value (tree_view->priv->hadjustment),
-		       gtk_tree_view_get_effective_header_height (tree_view));
-      gdk_window_move (tree_view->priv->header_window,
-		       - gtk_adjustment_get_value (tree_view->priv->hadjustment),
-		       0);
+
+      if (!tree_view->priv->in_top_row_to_dy)
+        {
+          gdk_window_move (tree_view->priv->bin_window,
+                  - gtk_adjustment_get_value (tree_view->priv->hadjustment),
+                  gtk_tree_view_get_effective_header_height (tree_view));
+          gdk_window_move (tree_view->priv->header_window,
+                  - gtk_adjustment_get_value (tree_view->priv->hadjustment),
+                  0);
+        }
       dy = tree_view->priv->dy - (int) gtk_adjustment_get_value (tree_view->priv->vadjustment);
       if (dy && tree_view->priv->edited_column)
         {
@@ -11313,9 +11316,12 @@ gtk_tree_view_adjustment_changed (GtkAdjustment *adjustment,
                     }
                 }
         }
-      tree_view->priv->in_scroll = TRUE;
-      gdk_window_scroll (tree_view->priv->bin_window, 0, dy);
-      tree_view->priv->in_scroll = FALSE;
+      if (!tree_view->priv->in_top_row_to_dy)
+        {
+          tree_view->priv->in_scroll = TRUE;
+          gdk_window_scroll (tree_view->priv->bin_window, 0, dy);
+          tree_view->priv->in_scroll = FALSE;
+        }
 
       if (dy != 0)
         {
-- 
1.9.2

